src/main/scala/com/fulcrumgenomics/bam/Bams.scala

```diff
@@ -0,0 +1,207 @@
+  def allSupplementaryAndSecondary: Iterator[SAMRecord] =
+    r1Supplementals.iterator ++ r2Supplementals.iterator ++ r1Secondaries.iterator ++ r2Secondaries.iterator
+
+  /** The total count of records for the template. */
+  def readCount: Int = r1.size + r2.size + r1Supplementals.size + r2Supplementals.size + r1Secondaries.size + r2Secondaries.size
+}
+
+object Template {
+  /**
+    * Generates a Template for the next template in the buffered iterator. Assumes that the
+    * iterator is queryname sorted or grouped.
+    */
+  def apply(recs: BetterBufferedIterator[SAMRecord]): Template = {
+    require(recs.hasNext)
+    val name = recs.head.getReadName
+    apply(recs.takeWhile(_.getReadName == name))
+  }
+
+
+  /** Generates a Template object from an iterator of SAMRecords from the same template. */
+  def apply(recs: Iterator[SAMRecord]): Template = {
+    var r1: Option[SAMRecord] = None
+    var r2: Option[SAMRecord] = None
+    val r1Supp = ListBuffer[SAMRecord]()
+    val r2Supp = ListBuffer[SAMRecord]()
+    val r1Sec  = ListBuffer[SAMRecord]()
+    val r2Sec  = ListBuffer[SAMRecord]()
+
+    recs foreach { r =>
+      val isR1 = !r.getReadPairedFlag || r.getFirstOfPairFlag
+      if (isR1) {
+        if      (r.getNotPrimaryAlignmentFlag) r1Sec += r
+        else if (r.getSupplementaryAlignmentFlag) r1Supp += r
+        else if (r1.isDefined) throw new IllegalArgumentException(s"Multiple non-secondary, non-supplemental R1s for ${r.getReadName}")
+        else r1 = Some(r)
+      }
+      else {
+        if      (r.getNotPrimaryAlignmentFlag)    r2Sec += r
+        else if (r.getSupplementaryAlignmentFlag) r2Supp += r
+        else if (r2.isDefined) throw new IllegalArgumentException(s"Multiple non-secondary, non-supplemental R2s for ${r.getReadName}")
+        else r2 = Some(r)
+      }
+    }
+
+    Template(r1=r1, r2=r2, r1Supplementals=r1Supp.toList, r2Supplementals=r2Supp.toList, r1Secondaries=r1Sec.toList, r2Secondaries=r2Sec.toList)
+  }
+}
+
+/**
+  * Utility methods for working with BAMs.
+  */
+object Bams extends LazyLogging {
+  /**
+    * Builds a sorting collection for sorting SAMRecords.
+    *
+    * @param order the desired output sort order
+    * @param header the header to use to encode records during sorting
+    * @param maxInMemory the maximum number of records to keep and sort in memory
+    * @param tmpDir an optional temp directory to use for temporary sorting files
+    * @return a SortingCollection
+    */
+  def sortingCollection(order: SortOrder,
+                        header: SAMFileHeader,
+                        maxInMemory: Int = 1e6.toInt,
+                        tmpDir: Option[DirPath] = None): SortingCollection[SAMRecord] = {
+    SortingCollection.newInstance(
+      classOf[SAMRecord],
+      new BAMRecordCodec(header),
+      if (order == SortOrder.queryname) new SAMRecordQueryNameComparator else new SAMRecordCoordinateComparator,
```

Comment:
They're the only ones in the spec ;)  But yes, for you, I'll use `order.getComparatorInstance()` instead.